internal/app/machined/pkg/controllers/network/cmdline.go

```diff
@@ -0,0 +1,122 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+package network
+
+import (
+	"fmt"
+	"net"
+	"sort"
+	"strings"
+
+	"github.com/talos-systems/go-procfs/procfs"
+	"inet.af/netaddr"
+)
+
+// CmdlineNetworking contains parsed cmdline networking settings.
+type CmdlineNetworking struct {
+	DHCP         bool
+	Address      netaddr.IPPrefix
+	Gateway      netaddr.IP
+	Hostname     string
+	LinkName     string
+	DNSAddresses []netaddr.IP
+	NTPAddresses []netaddr.IP
+}
+
+// ParseCmdlineNetwork parses `ip=` kernel cmdline argument producing all the available configuration options.
+//
+//nolint:gocyclo,cyclop
+func ParseCmdlineNetwork(cmdline *procfs.Cmdline) (settings CmdlineNetworking, err error) {
+	ipSettings := cmdline.Get("ip").First()
+	if ipSettings == nil {
+		return
+	}
+
+	// https://www.kernel.org/doc/Documentation/filesystems/nfs/nfsroot.txt
+	// ip=<client-ip>:<server-ip>:<gw-ip>:<netmask>:<hostname>:<device>:<autoconf>:<dns0-ip>:<dns1-ip>:<ntp0-ip>
+	fields := strings.Split(*ipSettings, ":")
+
+	// If dhcp is specified, we'll handle it as a normal discovered
+	// interface
+	if len(fields) == 1 && fields[0] == "dhcp" {
+		settings.DHCP = true
+	}
+
+	if !settings.DHCP {
+		for i := range fields {
+			if fields[i] == "" {
+				continue
+			}
+
+			switch i {
+			case 0:
+				settings.Address.IP, err = netaddr.ParseIP(fields[0])
+				if err != nil {
+					return settings, fmt.Errorf("cmdline address parse failure: %s", err)
+				}
+
+				// default is to have complete address masked
+				settings.Address.Bits = settings.Address.IP.BitLen()
+			case 2:
+				settings.Gateway, err = netaddr.ParseIP(fields[2])
+				if err != nil {
+					return settings, fmt.Errorf("cmdline gateway parse failure: %s", err)
+				}
+			case 3:
+				var netmask netaddr.IP
+
+				netmask, err = netaddr.ParseIP(fields[3])
+				if err != nil {
+					return settings, fmt.Errorf("cmdline netmask parse failure: %s", err)
+				}
+
+				ones, _ := net.IPMask(netmask.IPAddr().IP).Size()
+
+				settings.Address.Bits = uint8(ones)
+			case 4:
+				settings.Hostname = fields[4]
+			case 5:
+				settings.LinkName = fields[5]
+			case 7, 8:
+				var dnsIP netaddr.IP
+
+				dnsIP, err = netaddr.ParseIP(fields[i])
+				if err != nil {
+					return settings, fmt.Errorf("error parsing DNS IP: %w", err)
+				}
+
+				settings.DNSAddresses = append(settings.DNSAddresses, dnsIP)
+			case 9:
+				var ntpIP netaddr.IP
+
+				ntpIP, err = netaddr.ParseIP(fields[i])
+				if err != nil {
+					return settings, fmt.Errorf("error parsing DNS IP: %w", err)
+				}
+
+				settings.NTPAddresses = append(settings.NTPAddresses, ntpIP)
+			}
```

Comment:
What about `default:` that would return an error?

Reply:
I'm not sure what do you mean by that? `default:` is not valid, but empty value defaults to whatever DHCP provides (we don't need it here, as we only need static part)

Reply:
got it, we ignore extra params (as probably kernel does), so we protect ourselves from something which is parsed by the kernel, but not us